src/awkward/operations/ak_almost_equal.py

```diff
@@ -105,35 +105,89 @@ def visitor(left, right) -> bool:
+                    (left.dtype == right.dtype)
+                    and backend.nplike.all(left.data == right.data)
+                    and left.shape == right.shape
+                )
+            else:
+                return (
+                    is_approx_dtype(left.dtype, right.dtype)
+                    and backend.nplike.all(
+                        backend.nplike.isclose(
+                            left.data, right.data, rtol=rtol, atol=atol, equal_nan=False
+                        )
                     )
+                    and left.shape == right.shape
                 )
-                and left.shape == right.shape
-            )
-
         elif left.is_option:
             return backend.index_nplike.array_equal(
                 left.index.data < 0, right.index.data < 0
             ) and visitor(left.project().to_packed(), right.project().to_packed())
         elif left.is_union:
-            return (len(left.contents) == len(right.contents)) and all(
-                visitor(left.project(i).to_packed(), right.project(i).to_packed())
-                for i, _ in enumerate(left.contents)
+            # For two unions with different content orderings to match, the tags should be equal at each index
+            # Therefore, we can order the contents by index appearance
+            def ordered_unique_values(values):
+                # First, find unique values and their appearance (from smallest to largest)
+                # unique_index is in ascending order of `unique` value
+                (
+                    unique,
+                    unique_index,
+                    *_,
+                ) = backend.index_nplike.unique_all(values)
+                # Now re-order `unique` by order of appearance (`unique_index`)
+                return values[backend.index_nplike.sort(unique_index)]
+
+            # Find order of appearance for each union tags, and assume these are one-to-one maps
+            left_tag_order = ordered_unique_values(left.tags.data)
+            right_tag_order = ordered_unique_values(right.tags.data)
+
+            # Create map from left tags to right tags
+            left_tag_to_right_tag = backend.index_nplike.empty(
+                left_tag_order.size, dtype=np.int64
             )
+            left_tag_to_right_tag[left_tag_order] = right_tag_order
+
+            # Map left tags onto right, such that the result should equal right.tags
+            # if the two tag arrays are equivalent
+            new_left_tag = left_tag_to_right_tag[left.tags.data]
+            if not backend.index_nplike.all(new_left_tag == right.tags.data):
+                return False
+
+            # Now project out the contents, and check for equality
+            for i, j in zip(left_tag_order, right_tag_order):
+                if not visitor(
+                    left.project(i).to_packed(), right.project(j).to_packed()
+                ):
+                    return False
+            return True
+
```

Comment:
Previously, this logic did not accommodate unions with different orderings of `contents` (and correspondingly re-ordered `tags`).

Reply:
Does that mean that this is a bug-fix on a PR about a distinct new feature? (Ideally, they'd be separated, but in a pinch, combining them is okay.)

Reply:
A bugfix that this test suite uses. I'll move it into a new PR that sequences before this one.